### ISSUE TITLE
Adding (self-hosted) Firefox Sync to Bromite

### ISSUE THREAD

**Issue by makapao** [bromite/bromite#244]
**Is your feature request related to privacy?**
Yes.

**Is there a patch available for this feature somewhere?**
Not exactly an android-related patch, but there is some documentation from another project that might help.

**Describe the solution you'd like**

The ability to sync bookmarks, history, passwords and open tabs with Firefox.

At the moment there is no way for the user to share bookmarks, history or passwords with another browser, apart from manually importing bookmarks. Firefox Sync is a decent end-to-end encrypted option to share information between multiple browsers and an Android device on-the-go. 

Firefox Sync support would allow the end-to-end encrypted syncing of bookmarks, history, passwords, and open tabs with both desktop and mobile Firefox. For users of Firefox desktop it would make migrating to Bromite a more attractive prospect. For existing users it would add functionality.

**Describe alternatives you've considered**
I currently use either Firefox Notes (or Google Keep if I can't help it) to copy URLs and history I want to use between desktop Firefox on Linux and Bromite on Android. A lot more cumbersome, but it works.

**Other implementations**

**Epiphany (gnome browser)** have [added this feature](https://camo.githubusercontent.com/767d7f5b3f482217c4ce96203bfed8c0389e90d4/68747470733a2f2f692e696d6775722e636f6d2f57584a6374304a2e6a7067) and the following might be useful for understanding how they did it:

Description: [https://github.com/GNOME/epiphany/blob/70755404799e4393bf5605c9debec6877acf0177/lib/sync/README](https://blogs.gnome.org/mcatanzaro/2017/08/09/on-firefox-sync/)

Possibly related commit: [https://github.com/GNOME/epiphany/tree/70755404799e4393bf5605c9debec6877acf0177/lib/sync](https://github.com/GNOME/epiphany/tree/70755404799e4393bf5605c9debec6877acf0177/lib/sync)

Explainer on Epiphany Firefox Sync Support: [https://github.com/GNOME/epiphany/blob/70755404799e4393bf5605c9debec6877acf0177/lib/sync/README](https://github.com/GNOME/epiphany/blob/70755404799e4393bf5605c9debec6877acf0177/lib/sync/README)

**Further Reading**
Mozilla documentation

Mozilla host their sync client documentation here: [https://github.com/GNOME/epiphany/blob/70755404799e4393bf5605c9debec6877acf0177/lib/sync/README](https://moz-services-docs.readthedocs.io/en/latest/sync/)
There is also a range of documentation here for "for people who wish to understand the details of how Sync is implemented or wish to interact with it": [https://developer.mozilla.org/en-US/docs/Mozilla/Projects/Firefox_Sync](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/Firefox_Sync)

**Firefox Sync** on MDN (Mozilla Developer Network) Webdocs: [https://developer.mozilla.org/en-US/docs/Archive/Mozilla/Firefox_Sync](https://developer.mozilla.org/en-US/docs/Archive/Mozilla/Firefox_Sync)
**Sync Javascript Client API** on MDN (Mozilla Developer Network) Webdocs: [https://developer.mozilla.org/en-US/docs/Archive/Mozilla/Firefox_Sync/JavaScript_Client_API](https://developer.mozilla.org/en-US/docs/Archive/Mozilla/Firefox_Sync/JavaScript_Client_API)

[EDIT] Apologies for the multiple edits. Wanted to get this right.

**Comment by csagan5** [comment]
@legendavey this basically would add support for a cloud service (Firefox Sync), which is not federated. This is directly against the idea behind Bromite, which is to not use any cloud integration of any provider.

Furthermore, we are talking here about hundreds of hours of development to accomplish this. There is no way I would commit to such effort which is also out of line with the mission; feel free to implement it as a patch or a set of patches though.

**Comment by Bluscream** [comment]
Would a "patch" be enforced for all users or are they optional?

**Comment by csagan5** [comment]
@Bluscream a patch as needed for this issue would not be enabled by default, regardless.

**Comment by Offerel** [comment]
I know that's closed, but I would love such a feature. Currently im using 2 backends to sync my bookmarks. Firefox Sync and my own Firefox Addon. Both can work together. The later one sync to a personal WebDAV server as JSON file or alternatively to a PHP backend. 

This ensure that the bookmarks stay private and are not synced against a third party cloud. 

However since bookmark api isn't available for Firefox mobile, i can't extend the addon to work with Firefox mobile to. I don't know if there is way to import/export/sync bookmarks with bromite in any way, but if there is some way, the alternative backend would be usable also with bromite.

I didn't have any skills in Android coding, but I can change the backend to support requests from bromite, if bromite could do that. I could think of a trigger, if a bookmark is added, edited, moved or deleted. That trigger could send a simple XHR request to the PHP backend. Additionally we would need something, that 'pulls' changes from the backend, just in case the server has changes.

But that's only a rough idea. I don't know if bromite could do such things.

BTW: the backend work also with Chrome Desktop. I haven't released it in the Chrome Store since i didn't use Chrome. But a untested crx file is available in my repository. So in theory it should work.

**Comment by x11x** [comment]
I am also looking for a way to sync my browser data (bookmarks, tabs, etc). Enjoying using Bromite but this is the main feature for me that keeps me using Firefox instead.
Just wanted to outline how I think this should work in case anyone is interested in doing this. (I will experiment with it when I get time).
In my opinion, sync should be done through the Android ContentProvider system and kept out of Bromite.
In theory, a (3rd party) app should be able to query the ChromeBrowserProvider (after requesting permissions in the manifest) and get searches, bookmarks and history. Here is the implementation in chromium https://github.com/chromium/chromium/blob/master/chrome/android/java/src/org/chromium/chrome/browser/provider/ChromeBrowserProvider.java just to see what sort of information is there and how its accessed/provided.
I think the first step would be to do some experiments to make sure this data can be accessed from Bromite, and if any patches are to added to Bromite they should be to get the content provider working. [This StackOverflow answer](https://stackoverflow.com/questions/19826500/android-api-getting-browser-bookmarks-history) has a starting point for how this would be done in code. I guess you would have to change "com.android.chrome" to "org.bromite.bromite" in the URI?
In other words, I'm suggesting a separate app could be created (e.g. by @Offerel ) and released to e.g. F-Droid that simply queries the content provider in the background and e.g. sends to configurable WebDAV server. This would probably need to be done periodically using an AlarmManager (see https://developer.android.com/training/scheduling/alarms). Or there is a ContentObserver API, not sure if this is to be used by long running services though.
Anyone else could also implement their own sync system, e.g. to Firefox Sync, Owncloud, IPFS, whatever.

**Comment by csagan5** [comment]
@x11x sounds good, if it is really viable. Bromite is not going to partner with any cloud services for obvious reasons.

I also thought of some solution using WebDAV, but doing this with an external app is even better.

**Comment by jeslinmx** [comment]
> 
> 
> @legendavey this basically would add support for a cloud service (Firefox Sync), which is not federated. This is directly against the idea behind Bromite, which is to not use any cloud integration of any provider.
> 
> Furthermore, we are talking here about hundreds of hours of development to accomplish this. There is no way I would commit to such effort which is also out of line with the mission; feel free to implement it as a patch or a set of patches though.

To address the first point, it is actually possible to selfhost Firefox Sync, both the storage (https://mozilla-services.readthedocs.io/en/latest/howtos/run-sync-1.5.html) and the authentication (https://mozilla-services.readthedocs.io/en/latest/howtos/run-fxa.html) backends. I'm not sure if this qualifies it as "federated", though. Also, although this has been possible since 2014 (https://blog.mozilla.org/services/2014/05/08/firefox-accounts-sync-1-5-and-self-hosting/), and so _should_ be fairly mature, it seems that it is still a more involved process than it needs to be:

> This guide is preliminary and vastly incomplete.
- https://mozilla-services.readthedocs.io/en/latest/howtos/run-fxa.html

> The steps seem to be easy but it is not that easy.
- u/gullilike. https://www.reddit.com/r/selfhosted/comments/9eo56n/help_guide_own_firefox_sync_server_including_own/

I am interested in trying this out, so I'll post back on my results on whether it is easy to start up a Firefox backend.

And yes, I acknowledge the second point, in that this could draw hundreds of hours of development away for what is at best a user comfort feature, and if poorly implemented will in fact be an anti-feature. I am unable to contribute patches in any way, just wanted to put this out there to understand if the selfhostability of FxA would make it viable.

You guys rock, keep it up!

**Comment by csagan5** [comment]
> this could draw hundreds of hours of development away for what is at best a user comfort feature, and if poorly implemented will in fact be an anti-feature. I am unable to contribute patches in any way, just wanted to put this out there to understand if the selfhostability of FxA would make it viable.

I would be glad to accept a patch for this, if there is a maintainer willing to keep it up to date across releases. I might setup open source development bounties for this and other features; thanks for your report.

**Comment by csagan5** [comment]
Re-opening this feature to see if it viable to add support for at least part of it.

**Comment by uazo** [comment]
I make another proposal
once there was a test server (https://chromium.googlesource.com/chromium/chromium/+/trunk/sync/tools/testserver/chromiumsync.py) but it's died 6 years ago.
brave dudes are developing https://github.com/brave/go-sync, which is based on that protocol.
although it is not obvious why they do not have separate patches, I have seen that the changes are localized in the code

here their modifications to the chromium code
https://github.com/brave/brave-core/tree/master/chromium_src/components/sync

here the actual patch
https://github.com/brave/brave-core/tree/master/components/sync

they use the protocol that was originally used by the test server, so I guess little or no changes to the actual code that
then does the sync, brilliant.

among other things it seems to be the official protocol of google servers, but the server code seems not public

**Comment by csagan5** [comment]
That's very interesting!

**Comment by pixincreate** [comment]
Any improvements found.? I'd love to hear that.    
I've found that Brave sync works fine on devices.  
But the main caviet for me is that, it requires Brave to be installed on Mobile as well, which I don't want. As of now, I'm running Bromite on Android and Brave for my PC with KDE Connect for sending links in from Android to PC, but not vice-versa(sadly)

**Comment by foolab** [comment]
I spent the past few weeks tackling this from another angle.
I created a patch against pristine chromium 91.0.4466.5 that enables using a custom sync server.
I also created a minimal sync server from scratch that is -to the best of my abilities- compatible with the protocol.

The server seems to sustain updating and committing entries now but I am sure there are issues.
I don't yet handle deletion
I have no way to properly import session data from chrome (bookmarks can be dumped using the local sync backend and then imported to my custom server).
The code is not pretty
I use sqlite but I do not like it.
sometimes the server rejects some entries and the sync marker logic needs to be reworked.

This is the patch FYI https://gist.github.com/foolab/8c29f964a90334f340ddd8aced5a1aac
I can also share the server code if there is interest and/or someone to give me a hand.

**Comment by uazo** [comment]
basically you deleted the system account management and entered a ui to indicate an account.

in bromite the problem is that the synchronization must be re-enabled without re-enabling anything else so the various `isGooglePlayServicesAvailable` / `canUseGooglePlayServices` must be modified in the caller and not in the static method.
you also need to understand if by chance you are using gcm which is not present in bromite

I wonder why it was also necessary to modify `KidsManagementURLCheckerClient` which at first glance does not seem to have much to do with it and if you tried to use a brave server instance

**Comment by foolab** [comment]
> basically you deleted the system account management and entered a ui to indicate an account.
More or less yes. I added a basic UI to specify a custom sync server as well as an email and password for authentication.
I generate an auth token which is basically sha256(email:password). This avoids calls to google servers for authentication.

The patch then overrides the sync server url upon startup by passing --sync-url to chromium.

> in bromite the problem is that the synchronization must be re-enabled without re-enabling anything else so the various `isGooglePlayServicesAvailable` / `canUseGooglePlayServices` must be modified in the caller and not in the static method.
I tried not to be intrusive so I can rebase the patch in the future.
isGooglePlayServicesAvailable / canUseGooglePlayServices calls are part of chromium not gms.

> you also need to understand if by chance you are using gcm which is not present in bromite
No, I am not using it at all. I am using a phone without gms for testing.

> I wonder why it was also necessary to modify `KidsManagementURLCheckerClient` which at first glance does not seem to have much to do with it and if you tried to use a brave server instance

KidsManagementURLCheckerClient is a remains of me debugging sync failures.
I was returning true from SystemAccountManagerDelegate2.hasFeature() which was causing calls to that function. I should undo its modifications.

I have not tried using brave sync for 2 reasons:
1. The protocol is old
2. I wanted as minimal changes as possible.

**Comment by pixincreate** [comment]
> I generate an auth token which is basically sha256(email:password). This avoids calls to google servers for authentication.  
Well, I don't think using email and passqord would be optimal/private unless you use a mail which is something like ProtonMail.  
Instead of that, we can follow **Brave** in sync as it doesn't require any type of account but a ```pass-phrase``` that is of few words(_30?_).   
I know that this is not an easy task, but if viewing in terms of user's privacy, Brave's approach seems to hold good.

**Comment by uazo** [comment]
> I tried not to be intrusive so I can rebase the patch in the future.

well, for bromite changing `isGooglePlayServicesAvailable` to true is highly intrusive!
a patch made like this is difficult to acquire (but here I decide nothing), and does not make it clear that it is the minimum to be activated to make it work, and therefore difficult to maintain.

You should look for another approach, which is to set them to false and see what needs to be enabled for sync to work, or you already tried?

> The protocol is old

can I ask you for more details?

**Comment by foolab** [comment]
> > I generate an auth token which is basically sha256(email:password). This avoids calls to google servers for authentication.
> 
> Well, I don't think using _email and password_ would be optimal/private unless you use a mail which is something like ProtonMail that has encryption.

You can use ceo@google.com as your email. It does not matter. The email itself is not used for anything other than generating the "fake token". It does not get sent to google servers AFAICT.

> Instead of that, we can follow **Brave** in sync as it doesn't require any type of account but a `pass-phrase` that is of few words(_30?_).
> I know that this is not an easy task, but if viewing in terms of user's privacy, Brave's approach seems to hold good.

It is an easy task as I control both the server and the client. I cooked my own python server to handle the sync. I just need to push it somewhere.

> > I tried not to be intrusive so I can rebase the patch in the future.
> 
> well, for bromite changing `isGooglePlayServicesAvailable` to true is highly intrusive!
> a patch made like this is difficult to acquire (but here I decide nothing), and does not make it clear that it is the minimum to be activated to make it work, and therefore difficult to maintain.

I understand. I am not proposing taking my patch. I am just proposing a general direction that can be taken. The approach I am suggesting is:
1. Use the existing sync code in chromium
2. override the authentication and token generation for accounts
3. override the url for the sync server
4. implement a custom sync server
5. profit.
The reason behind point 3 is to allow any user to host his own sync server to keep his data private.

> You should look for another approach, which is to set them to false and see what needs to be enabled for sync to work, or you already tried?

I really do not remember. I am sure it's doable at the expense of a larger patch :-)

> 
> > The protocol is old
> 
> can I ask you for more details?

Seems brave protocol is based on an old revision of the actual google sync protocol (https://github.com/brave/go-sync/blob/master/schema/protobuf/sync_pb/sync.proto#L770) seems to be revision 52 (I am not sure) while we are currently using revision 90.
Using an old protocol is not a problem in itself. The only issue is we will have to maintain it ourselves (or constantly sync from brave).
I have no idea how large the brave patchset is but I am sure it's larger than what I did ;-)

To summarize:
1. We either use sync code from brave and sync server from brave and hope bromite can continue to use brave patches OR
2. We use the pristine chromium sync code and implement a minimal custom server (Which is what I did).

**Comment by csagan5** [comment]
> * We either use sync code from brave and sync server from brave and hope bromite can continue to use brave patches OR
> * We use the pristine chromium sync code and implement a minimal custom server (Which is what I did).

None of these two adds firefox sync to Bromite (which is the issue at hand here); some other solution has been proposed in previous comments as well.

> a patch made like this is difficult to acquire (but here I decide nothing)

Me neither: this decision makes itself: I cannot create or maintain a patch for this feature, so the only option is that someone builds it and maintains it across releases. See [my previous comment](https://github.com/bromite/bromite/issues/244#issuecomment-515672632):

> I would be glad to accept a patch for this, if there is a maintainer willing to keep it up to date across releases

The only criteria for adding it are:
* does not introduce privacy or security problems
* does not have obvious bugs that users will report since day 1

**Comment by pixincreate** [comment]
I believe that adding sync to Brave would be easier than Firefox.  
But, the caveat for us now is that, Bromite, being a chromium based browser, will find it hard to sync with firefox(not sure).  
However, on the PC side, I'd recommend making an extension for chromium as well as gecko, which requires DBUS to be setup on windows.  

> None of these two adds firefox sync to Bromite (which is the issue at hand here); some other solution has been proposed in previous comments as well.   

AFAIK, you need Firefox account to be able to sync to sync with Firefox browsers. In that case, all that we need to do is to setup an environment for Bromite to add Firefox account as the server exists already.   
(correct me if I'm wrong).

**Comment by uazo** [comment]
> that someone builds it and maintains it across releases.

@foolab  if you want to try it, I am available for support and reviews.

**Comment by foolab** [comment]
> > that someone builds it and maintains it across releases.
> 
> @foolab if you want to try it, I am available for support and reviews.

@uazo I can try porting it to bromite and I can try fixing the isGooglePlayServicesAvailable / canUseGooglePlayServices part but unfortunately I am unable to maintain it across releases.

I also see that the general direction for bromite is to use firefox sync which is not something I am personally interested in right now :-(

**Comment by kokukyocho** [comment]
Hello @makapao!

I am a bot :robot:

Thanks for submitting this issue!
I noticed that it is missing the template, please edit the issue to match the template for either a [bug](https://github.com/bromite/bromite/blob/master/.github/ISSUE_TEMPLATE/bug_report.md) or a [feature request](https://github.com/bromite/bromite/blob/master/.github/ISSUE_TEMPLATE/feature_request.md).

The issue will be automatically re-opened after that.

**Tip**: if you use GitHub in Desktop mode the template will be proposed to you when submitting issues.

**Comment by csagan5** [comment]
> But, the caveat for us now is that, Bromite, being a chromium based browser, will find it hard to sync with firefox(not sure).

There is no sync with Firefox (Mozilla), the only code used would be the client side and the self-hosted server side one, see [this comment](https://github.com/bromite/bromite/issues/244#issuecomment-514509401). It will never sync with a third-party cloud service.

> AFAIK, you need Firefox account to be able to sync to sync with Firefox browsers. In that case, all that we need to do is to setup an environment for Bromite to add Firefox account as the server exists already.

This is undesirable. Once this feature is added one can choose the backend service (Firefox sync, or WebDAV, the latter being preferable because an open standard); if you have no self-hosted or otherwise trusted backend service you would not be able to use this feature.

> I also see that the general direction for bromite is to use firefox sync

There is no "general direction", I already wrote down the criteria, let me extend it: the patch to implement this feature should:
* not introduce privacy or security problems
* not have obvious bugs that users will report since day 1
* have a maintainer that will take care of bug fixes and porting it across major versions
* there must be an open source (working) backend project available so that anyone can run it

I know it is ambitious, but anything else would not be in line with the goals of this project.

**Comment by Offerel** [comment]
If there would be support for We extensions and the bookmark api, you can use my AddOn for syncing bookmarks with WebDAV or alternatively a selhosted PHP backend. Have a look for Syncmarks in my repositories. This can be used with Chromium and Firefox.

**Comment by Goldmaster** [comment]
This would be very awesome. As according to privacy guides, firefox is get to support a few features on mobile and bromite is seen as a good app. Firefox on desktop is recommended.

So adding option to sync with firefox desktop would be alwsome opportunity here

https://privacyguides.org/browsers/#mobile-only-recommendations

**Comment by uazo** [comment]
> So adding option to sync with firefox desktop would be alwsome opportunity here

I think the priority goes to Bromite Desktop

**Comment by saxofooon** [comment]
What about [xbrowsersync](https://www.xbrowsersync.org/)? Cannot install it on /e/OS at the moment.

**Comment by csagan5** [comment]
> > So adding option to sync with firefox desktop would be alwsome opportunity here
> 
> I think the priority goes to Bromite Desktop

There are no plans to develop Bromite for desktop and there is already ungoogled-chromium for desktop.

> https://privacyguides.org/browsers/#mobile-only-recommendations
> These features are supported in Bromite as it uses Chromium WebView which is included in all Android operating systems. 

This is incorrect, Bromite does not use the Chromium WebView.

**Comment by cjwijtmans** [comment]
personally what i am looking for is local syncing without self hosting a central server. Something like brave sync but without going over the internet and sync on local network only.

**Comment by HammyHavoc** [comment]
In Tachiyomi on Android (a comic reading app), I was able to choose a Nextcloud app folder as my target dir for backups, which then get synced to my server via the Nextcloud app automatically.

On that train of thought, what might make more sense is treating browser data as flat files and having a file per device, e.g., `HAMMYSTHINKPAD`, `HAMMYSPHONE`, and `HAMMYSDESKTOPPC` in a dir that I control and can utilize in whatever way makes sense. It's simple, it's crude, but perhaps that's exactly what it needs to be as this data is already getting saved locally anyway.

**Comment by tajnymag** [comment]
> In Tachiyomi on Android (a comic reading app), I was able to choose a Nextcloud app folder as my target dir for backups, which then get synced to my server via the Nextcloud app automatically.
> 
> On that train of thought, what might make more sense is treating browser data as flat files and having a file per device, e.g., `HAMMYSTHINKPAD`, `HAMMYSPHONE`, and `HAMMYSDESKTOPPC` in a dir that I control and can utilize in whatever way makes sense. It's simple, it's crude, but perhaps that's exactly what it needs to be as this data is already getting saved locally anyway.

That sounds quite insecure. If Bromite stored all userdata in a directory readable by other apps, it would be trivial to extract confidential information out of it. The data would have to be encrypted, which would bring another layer of complexity.

**Comment by HammyHavoc** [comment]
> > In Tachiyomi on Android (a comic reading app), I was able to choose a Nextcloud app folder as my target dir for backups, which then get synced to my server via the Nextcloud app automatically.
> > On that train of thought, what might make more sense is treating browser data as flat files and having a file per device, e.g., `HAMMYSTHINKPAD`, `HAMMYSPHONE`, and `HAMMYSDESKTOPPC` in a dir that I control and can utilize in whatever way makes sense. It's simple, it's crude, but perhaps that's exactly what it needs to be as this data is already getting saved locally anyway.
> 
> That sounds quite insecure. If Bromite stored all userdata in a directory readable by other apps, it would be trivial to extract confidential information out of it. The data would have to be encrypted, which would bring another layer of complexity.

People punch in a key on Authy on each device to decrypt their data. Also depends on what people want to have sync too, tabs and bookmarks are a far cry from passwords. For passwords, using Bitwarden in its own app makes more sense. But yes, it should be encrypted.

**Comment by SuperCowProducts** [comment]
Any news? Is there some fork of Bromite that has started experimenting this?
